### PR TITLE
chore: adding upgrade handler for 09-localhost removal 

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

- Adding upgrade handler for pruning of `09-localhost` client and consensus states
- Adding migration docs 

closes: #1201 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Comment by damiannolan** [comment]
I think this PR also requires some migration docs added to instruct devs to wire up the upgrade handler.
I will add some! 

I've added a new directory `upgrades/v5` under `core`. I am open to suggestions on moving this to somewhere else but I think this is a nice pattern to use for upgrade handlers for ibc core logic. Inspired from osmosis [here](https://github.com/osmosis-labs/osmosis/tree/main/app/upgrades)

File: docs/migrations/v4-to-v5.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+The `09-localhost` light client implementation has been removed because it is currently non-functional.
+
+An upgrade handler has been added to supply chain developers with the logic needed to prune the ibc client store and successfully complete the removal of `09-localhost`.
+Add the following to the application upgrade handler in `app/app.go`, calling `UpgradeToV5` to perform store migration logic.
+
+```go
+import (
+    // ...
+    ibcv5 "github.com/cosmos/ibc-go/v3/modules/core/upgrades/v5"
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply docs/migrations/v4-to-v5.md:68]: Correct!

File: docs/migrations/v4-to-v5.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+### Client Keeper
+
+Keeper function `CheckMisbehaviourAndUpdateState` has been removed since function `UpdateClient` can now handle updating `ClientState` on `ClientMessage` type which can be any `Misbehaviour` implementations.  
+
+### SDK Message
+
+`MsgSubmitMisbehaviour` is deprecated since `MsgUpdateClient` can now submit a `ClientMessage` type which can be any `Misbehaviour` implementations.
+
+The field `header` in `MsgUpdateClient` has been renamed to `client message`.
```
Comment (crodriguezvega) [inline docs/migrations/v4-to-v5.md:92]:
```suggestion
The field `header` in `MsgUpdateClient` has been renamed to `client_message`.
```

**Review by crodriguezvega — COMMENTED** [review]
Maybe we could call the folder `migrations`? Not very strongly opposed to `upgrades` anyway, but thought that `migrations` is how the folder is called in `docs` and we could also have migrations that not necessarily need to run on the upgrade to a major version. Well, just an idea!

File: docs/migrations/v4-to-v5.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+### `ConsensusState`
+
+The `GetRoot` function has been removed from consensus state interface since it was not used by core IBC.
+
+### Light client implementations
+
+The `09-localhost` light client implementation has been removed because it is currently non-functional.
+
+An upgrade handler has been added to supply chain developers with the logic needed to prune the ibc client store and successfully complete the removal of `09-localhost`.
```
Comment (colin-axner) [inline docs/migrations/v4-to-v5.md:62]: Might be good to specify this is only needed if the chain has a localhost created. Maybe we could provide a cli command for them to use to check?

File: modules/core/upgrades/v5/upgrades.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/cosmos/ibc-go/v3/modules/core/exported"
+)
+
+// Localhost is the client type for a localhost client. It is also used as the clientID
+// for the localhost client.
+const Localhost string = "09-localhost"
+
+// UpgradeToV5 prunes the 09-Localhost client and associated consensus states from the ibc store
+func UpgradeToV5(ctx sdk.Context, clientKeeper clientkeeper.Keeper) {
+	clientStore := clientKeeper.ClientStore(ctx, Localhost)
```
Comment (colin-axner) [inline modules/core/upgrades/v5/upgrades.go:20]: I wonder if there is a way to wipe the entire client store? I think you could iterate over just the client prefixed key, but I kinda like explicitly deleting consensus/client states

**Review by colin-axner — APPROVED** [review]
Looks great :) 

Might be nice to add information to the localhost removal changelog entry that this function can be used to remove existing localhost implementations

**Comment by colin-axner** [comment]
> Which approach should we take with naming? 

I like migrations. Typically this functionality would be run by SDK's `RunMigrations` function. It isn't only because most chains don't have a localhost thus making the extra computation unnecessary. I also want to avoid confusion with client/connection/channel upgrades

**Comment by damiannolan** [comment]
> > Which approach should we take with naming?
> 
> I like migrations. Typically this functionality would be run by SDK's `RunMigrations` function. It isn't only because most chains don't have a localhost thus making the extra computation unnecessary. I also want to avoid confusion with client/connection/channel upgrades

Sounds good, that's a compelling enough argument for me! I'll update this today with the naming changes throughout 👍

File: CHANGELOG.md [hunk 5]
```diff
@@ -57,7 +57,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * (modules/light-clients/07-tendermint) [\#1118](https://github.com/cosmos/ibc-go/pull/1118) Deprecating `AllowUpdateAfterExpiry and AllowUpdateAfterMisbehaviour`. See ADR-026 for context.
 * (modules/core/02-client) [\#1188](https://github.com/cosmos/ibc-go/pull/1188/files) Routing `MsgSubmitMisbehaviour` to `UpdateClient` keeper function. Deprecating `SubmitMisbehaviour` endpoint.
 * (modules/core/02-client) [\#1208](https://github.com/cosmos/ibc-go/pull/1208) Replace `CheckHeaderAndUpdateState` usage in 02-client with calls to `VerifyClientMessage`, `CheckForMisbehaviour`, `UpdateStateOnMisbehaviour` and `UpdateState`. 
-* (modules/light-clients/09-localhost) [\#1187](https://github.com/cosmos/ibc-go/pull/1187/) Removing localhost light client implementation as it is not functional. 
+* (modules/light-clients/09-localhost) [\#1187](https://github.com/cosmos/ibc-go/pull/1187/) Removing localhost light client implementation as it is not functional. An upgrade handler been provided in `modules/migrations/v5` to prune `09-localhost` clients and consensus states from the store.
```
Comment (crodriguezvega) [inline CHANGELOG.md:60]:
```suggestion
* (modules/light-clients/09-localhost) [\#1187](https://github.com/cosmos/ibc-go/pull/1187/) Removing localhost light client implementation as it is not functional. An upgrade handler is provided in `modules/migrations/v5` to prune `09-localhost` clients and consensus states from the store.
```

File: modules/core/migrations/v5/migrations.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	host "github.com/cosmos/ibc-go/v3/modules/core/24-host"
+	"github.com/cosmos/ibc-go/v3/modules/core/exported"
+)
+
+// Localhost is the client type for a localhost client. It is also used as the clientID
+// for the localhost client.
+const Localhost string = "09-localhost"
+
+// MigrateToV5 prunes the 09-Localhost client and associated consensus states from the ibc store
+func MigrateToV5(ctx sdk.Context, clientKeeper clientkeeper.Keeper) {
```
Comment (charleenfei) [inline modules/core/migrations/v5/migrations.go:19]: nit: we could potentially rename this function to make it more explicit as to what it does? (something like `Migrate_RemoveLocalHostClient`)

**Review by crodriguezvega — APPROVED** [review]
Just a small nit..
Nice work, @damiannolan!

File: docs/migrations/v4-to-v5.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+```go
+import (
+    // ...
+    ibcv5 "github.com/cosmos/ibc-go/v5/modules/core/migrations/v5"
+)
+
+// ...
+
+app.UpgradeKeeper.SetUpgradeHandler(
+    upgradeName,
```
Comment (charleenfei) [inline docs/migrations/v4-to-v5.md:74]: nit: `upgradeName` can be `MigrateToV5` or `MigrateLocalHostClient`?

**Review by charleenfei — APPROVED** [review]
nice :)

**Pull request merged**